src/sugarblas.jl

```diff
@@ -17,25 +16,24 @@ import Base.LinAlg.BLAS: blascopy!, scal!, scal
     DY
 end
 
-# Add robustness to scal!
-@inbounds function scal!(n::Integer, DA::Number, DX::AbstractArray, incx::Integer)
-    i = 1
-    n = min(n,length(DX))
-    @simd for i in 1:incx:n
-        DX[i] *= DA
-    end
-    DX
+function scal(n::Integer, DA::Number, DX::AbstractArray, incx::Integer)
+    O = copy(DX)
+    scal!(n, DA, O, incx)
 end
 
-# Add robustness to scal
-@inbounds function scal{T}(n::Integer, DA::Number, DX::AbstractArray{T}, incx::Integer)
-    zeros(T, size(DX)...)
+@inbounds function scal!(n::Integer, DA::Number, DX::AbstractArray, incx::Integer)
     i = 1
     n = min(n,length(DX))
-    @simd for i in 1:incx:n
-        O[i] = DA*DX[i]
+    if incx == 1
+        @simd for i in 1:n
+            DX[i] *= DA
```

Comment:
Before rapid firing another commit, I hope this does indeed work (already did the check with `@inbounds if then else` just in case), I'll open a PR for the `whole function` thing which is a bit confusing. I'm a bit embarrassed  it took me this many commits to place a macro.

```julia
function blascopy!(n::Number, DX::AbstractArray, incx::Integer, DY::AbstractArray, incy::Integer)
    x = 1
    y = 1
    nx = min(n, length(DX))
    ny = min(n, length(DY))
    @inbounds while x <= nx && y <= ny
        DY[y] = DX[x]
        x += incx
        y += incy
    end
    DY
end

function scal!(n::Integer, DA::Number, DX::AbstractArray, incx::Integer)
    i = 1
    n = min(n,length(DX))
    @inbounds if incx == 1
        @simd for i in 1:n
            DX[i] *= DA
        end
    else
        for i in 1:incx:n
            DX[i] *= DA
        end
    end
    DX
end
```